pygeo/parameterization/DVGeoCST.py

```diff
@@ -0,0 +1,1093 @@
+    comm : MPI communicator, optional
+        Communicator for DVGeometryCST instance, by default MPI.COMM_WORLD
+    isComplex : bool, optional
+        Initialize variables to complex types where necessary, by default False
+    debug : bool, optional
+        Show plots when addPointSet is called to visually verify that it is correctly splitting
+        the upper and lower surfaces of the airfoil points, by default False
+    """
+
+    def __init__(self, datFile, idxChord=0, idxVertical=1, comm=MPI.COMM_WORLD, isComplex=False, debug=False):
+        super().__init__(datFile)
+        self.xIdx = idxChord
+        self.yIdx = idxVertical
+        self.comm = comm
+        self.isComplex = isComplex
+        if isComplex:
+            self.dtype = complex
+            self.dtypeMPI = MPI.DOUBLE_COMPLEX
+        else:
+            self.dtype = float
+            self.dtypeMPI = MPI.DOUBLE
+        self.debug = debug
+        if debug and not pltImport:
+            raise ImportError("matplotlib.pyplot could not be imported and is required for DVGeoCST debug mode")
+
+        # Store the DVs and flags to determine if the limited options have already been specified
+        self.DVs = {}
+        self.DVExists = {
+            "upper": False,
+            "lower": False,
+            "n1_upper": False,
+            "n2_upper": False,
+            "n1_lower": False,
+            "n2_lower": False,
+            "chord": False,
+        }
+
+        # Default DVs to be copied for each point set
+        self.defaultDV = {
+            "upper": 0.1 * np.ones(8, dtype=self.dtype),
+            "lower": -0.1 * np.ones(8, dtype=self.dtype),
+            "n1_upper": np.array([0.5], dtype=self.dtype),
+            "n2_upper": np.array([1.0], dtype=self.dtype),
+            "n1_lower": np.array([0.5], dtype=self.dtype),
+            "n2_lower": np.array([1.0], dtype=self.dtype),
+            "n1": np.array([0.5], dtype=self.dtype),
+            "n2": np.array([1.0], dtype=self.dtype),
+            "chord": np.array([1.0], dtype=self.dtype),
+        }
+
+        # ========== Process the input airfoil and set variables accordingly ==========
+        coords = readCoordFile(datFile)
+        self.foilCoords = np.zeros_like(coords, dtype=self.dtype)
+        self.foilCoords[:, self.xIdx] = coords[:, 0]
+        self.foilCoords[:, self.yIdx] = coords[:, 1]
+
+        # Set the leading and trailing edge x coordinates
+        self.xMin = np.min(self.foilCoords[:, self.xIdx])
+        self.xMax = np.max(self.foilCoords[:, self.xIdx])
+
+        # Check that the leading edge is at y = 0
+        idxLE = np.argmin(self.foilCoords[:, self.xIdx])
+        yLE = self.foilCoords[idxLE, self.yIdx]
+        if abs(yLE) > 1e-2:
+            raise ValueError(f"Leading edge y (or idxVertical) value must equal zero, not {yLE}")
+
+        # Determine the trailing edge thickness
```

Comment:
Haven't looked into the code here, but it sounds [prefoil](https://github.com/mdolab/prefoil) might have some of the machinery/logic that would help here

Reply:
I think it would be nice to be able to support closed trailing edges (makes ADflow meshes easier), but I understand the problem with this approach. I'm open to implementing the angle detection approach. I don't see any tools to do this in prefoil but maybe I'm not looking hard enough; @marcomangano do you know where it'd be?

Reply:
[This method](https://github.com/mdolab/prefoil/blob/c59f087134517f2d4a9f1a621284da1f3abf7a67/prefoil/airfoil.py#L898) does the check that Anil is recommending AND opens up the airfoil. Not sure if it makes sense to have prefoil as a dependency for the cst module.

Reply:
I think @akleb worked on that most recently, if I recall correctly the check works on `coords` which is the point cloud and not the spline (`self.spline.X` are the actual points, I believe) - which is just re-generated at the end . Anyway, the idea should hold and should be quick to re-implement. Good luck!

Reply:
_\*gasp\*_ Did we just agree to code duplication in pygeo? Why not just pop the function that detects the trailing edge points out of the `removeTE` function into a new function and then have the `removeTE` function call it? Then you could just use the already written code. I don't really care either way, just seems like bad practice to rewrite code in different modules unless pygeo doesn't have a prefoil/pyspline dependency (the real argument for me would be if it didn't have a pyspline dependency (I know nothing about pygeo so idk what the case is)).

Also I am not really sure what is going on in this code, but perhaps there is a way to use more of prefoil if you're not already. I can talk with you about what it does, if you would like (I have no experience with DV objects).

Reply:
Anil and I are meeting later today to figure out the best approach since it is complicated by the fact that the trailing edge needs to be separated out for each partitioned group of points in the point set individually, rather than all airfoil coordinates at once. We'll keep prefoil in mind, thanks for the suggestions.